In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression

In [2]:
X , y = make_regression(n_samples=1000 , n_features=3 , noise=10 , random_state=42)

In [3]:
print(X.shape)

(1000, 3)


In [4]:
print(y.shape)

(1000,)


In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
def calculate_loss(X, y, weights):
    
    #  Calculate predictions by multiplying inputs (X) with weights:)
    predictions = X.dot(weights)
    
    #  Calculate the errors (difference between predictions and actual values):)
    errors = predictions - y
    
    # Calculate Mean Squared Error (MSE):)
    mse = np.mean(errors ** 2)
    
    #  Divide by 2 to match the gradient calculation format:)
    loss = mse / 2
    
    return loss

Gradient Descent

In [7]:
def gradient_descent(X, y, num_iterations , freq_print):
    y = y.reshape(len(y),1)
    alpha = 0.001
    X_new = np.c_[np.ones(len(X)), X]
    W = np.random.randn(X_new.shape[1], 1)
    for i in range(num_iterations) :    
        yhat = X_new.dot(W)
        error = yhat - y
        gradient = X_new.T.dot(error)
        W = W - alpha * gradient / len(X)
        if i % freq_print == 0:
            print(calculate_loss(X_new , y , W))
        
        
    return W             

In [8]:
W = gradient_descent(X_test , y_test , 10000 , 1000)

8305.39264944929
1215.6250337228785
228.64705145874316
78.99658841973812
54.413644754457856
50.085767410810604
49.28061703773948
49.12458062286925
49.093463951833364
49.087138310512636


In [9]:
W 

array([[ 1.60795158],
       [98.19389296],
       [81.98373753],
       [25.92503577]])

To test the gradient descent

this function give the most efficient weight and I print the loss for this weight

In [10]:
def least_weight(X, Y):
    Y = Y.reshape(len(Y), 1)
    X = np.c_[np.ones(X.shape[0]), X]
    X_transpose = X.T
    W = np.linalg.inv(X_transpose @ X) @ X_transpose @ Y
    
    
    print(calculate_loss(X , Y , W))
    

print the output for the test data withe the most efficient weights

In [11]:
least_weight(X_test,y_test)

49.08549457182744


The gradient descent final loss output is almost equal to the least loss output.......................
                                   49.087106110189694  |||  49.08549457182744